### PR TITLE
Add auto trainer

### PR DESCRIPTION
**Pull request by AndSt**

### CODE REVIEW

File: knodle/trainer/config.py [hunk 1]
```diff
@@ -14,11 +14,11 @@ class TrainerConfig:
     def __init__(
             self,
             criterion: Callable[[Tensor, Tensor], float] = cross_entropy_with_probs,
-            batch_size: int = 32,
+            batch_size: int = 256,
```
Comment (agsedova) [inline knodle/trainer/config.py:17]: isn't it too big default value? mb 64/128?

File: tutorials/baseline/data/imdb/keywords/negative-words.txt [hunk 2]
```diff
@@ -0,0 +1,4814 @@
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
```
Comment (agsedova) [inline tutorials/baseline/data/imdb/keywords/negative-words.txt:1]: don't think it is a good idea to have these files in github. mb download them from minio as others?

File: tutorials/baseline/auto_trainer.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+from sklearn.feature_extraction.text import TfidfVectorizer
+
+from knodle.model.logistic_regression_model import LogisticRegressionModel
+from knodle.trainer.auto_trainer import AutoTrainer
+
+# This python script contains rarely any explanation. For more description, we refer to the corresponding
+# jupyter notebook. There the steps are explained in more detail
+
+
+# Define some functions
```
Comment (agsedova) [inline tutorials/baseline/auto_trainer.py:25]: some small function descriptions are still lacking to me

File: tutorials/baseline/auto_trainer.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+def np_to_tensor_dataset(x: np.ndarray):
+    if isinstance(x, sp.csr_matrix):
+        x = x.toarray()
+    x = torch.from_numpy(x)
+    x = TensorDataset(x)
+    return x
+
+
+def create_tfidf_dataset(
```
Comment (agsedova) [inline tutorials/baseline/auto_trainer.py:40]: just to be sure: this function will be moved to transforemer.py in later refactoring?

File: tutorials/baseline/auto_trainer.py [hunk 5]
```diff
@@ -0,0 +1,111 @@
+import os
```
Comment (agsedova) [inline tutorials/baseline/auto_trainer.py:1]: there is no tutorial folder, and also I doubt it should be in baseline folder
I would suggest putting it to examples/trainer folder

File: tutorials/baseline/auto_trainer_example.ipynb [hunk 6]
```diff
@@ -0,0 +1,571 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "treated-mining",
+   "metadata": {},
+   "source": [
+    "## Further readings\n",
+    "\n",
+    "1. The next tutorial would be %TODO\n",
```
Comment (agsedova) [inline tutorials/baseline/auto_trainer_example.ipynb:543]: todo ?

File: tutorials/baseline/auto_trainer_example.ipynb [hunk 7]
```diff
@@ -0,0 +1,571 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "id": "declared-hierarchy",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "## TODO: Data Description"
```
Comment (agsedova) [inline tutorials/baseline/auto_trainer_example.ipynb:94]: todo?

File: tutorials/baseline/auto_trainer_example.ipynb [hunk 8]
```diff
@@ -0,0 +1,571 @@
...
+   "id": "obvious-chicago",
+   "metadata": {},
+   "source": [
+    "# Auto Trainer Tutorial\n",
+    "\n",
+    "Here we want to show you an maximally easy example how to use Knodle out-of-the box. Therefor this tutorial contains\n",
+    "the following steps:\n",
+    "1. Download from Knodle Server and load into memory\n",
+    "2. Initialize model and data for a TF_IDF and Logistic Regression example\n",
+    "3. Use AutoTrainer to use 2 weak training strategies\n",
```
Comment (agsedova) [inline tutorials/baseline/auto_trainer_example.ipynb:14]: .. but now in the tutorial, there is only a majority vote example, am I wrong? 
I think the idea of showing 2 weak training strategies is good, e.g.:
```
maj_trainer = AutoTrainer(
    name="majority", "knn", "dscrossweigh"
...
)
```

File: tutorials/baseline/auto_trainer_example.ipynb [hunk 9]
```diff
@@ -0,0 +1,571 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "id": "obvious-chicago",
+   "metadata": {},
+   "source": [
+    "# Auto Trainer Tutorial\n",
+    "\n",
+    "Here we want to show you an maximally easy example how to use Knodle out-of-the box. Therefor this tutorial contains\n",
```
Comment (agsedova) [inline tutorials/baseline/auto_trainer_example.ipynb:10]: could you add a little bit more detailed explanation of what autotrainer is? what is the difference between autotrainer and other (denoising specific) trainers?

File: tutorials/baseline/auto_trainer_example.ipynb [hunk 10]
```diff
@@ -0,0 +1,571 @@
...
+    "print(f\"Train Z dimension: {train_rule_matches_z.shape}\")\n",
+    "print(f\"Train avg. matches per sample: {train_rule_matches_z.sum() / train_rule_matches_z.shape[0]}\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "included-reservation",
+   "metadata": {},
+   "source": [
+    "### Preprocess data to TF_IDF values"
```
Comment (agsedova) [inline tutorials/baseline/auto_trainer_example.ipynb:222]: a couple of lines, why the data should be preprocessed to TF_IDF values?

**Review by agsedova — CHANGES_REQUESTED** [review]

File: tutorials/baseline/auto_trainer.py (hunk 5 above)
Comment (AndSt) [reply tutorials/baseline/auto_trainer.py:1]: Ah thanks. I'll have to investigate how that could happen ..

File: tutorials/baseline/auto_trainer.py (hunk 3 above)
Comment (AndSt) [reply tutorials/baseline/auto_trainer.py:25]: From above: "This python script contains rarely any explanation. For more description, we refer to the corresponding jupyter notebook. There the steps are explained in more detail."

File: tutorials/baseline/auto_trainer.py (hunk 4 above)
Comment (AndSt) [reply tutorials/baseline/auto_trainer.py:40]: yes

File: tutorials/baseline/auto_trainer_example.ipynb (hunk 6 above)
Comment (AndSt) [reply tutorials/baseline/auto_trainer_example.ipynb:543]: Good catch. CW should be there

File: tutorials/baseline/auto_trainer.py (hunk 3 above)
Comment (agsedova) [reply tutorials/baseline/auto_trainer.py:25]: yes, I have seen it, and it is ok that there is no detailed explanation - but I think one-line comments for the next two functions would be good (as it is for create_tfidf_dataset function). But not so important actually :)

Comment (AndSt) [reply tutorials/baseline/auto_trainer.py:25]: I'm not sure where. np_to_tensor_dataset should be rather descriptive?

File: tutorials/baseline/auto_trainer_example.ipynb (hunk 8 above)
Comment (AndSt) [reply tutorials/baseline/auto_trainer_example.ipynb:14]: Not sure if we discussed it's part of my PR

File: examples/trainer/simple_auto_trainer/notebook.ipynb [hunk 11]
```diff
@@ -0,0 +1,303 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "operational-presence",
+   "metadata": {},
+   "source": [
+    "### Data description\n",
+    "\n",
+    "We have three splits: train, develop and test split. For each split, there is DataFrame, holding text, and a Z matrix, relating instances, or rows in the DataFrame, to rules. Again, for more information we refer to the creation of the Dataset https://github.com/knodle/knodle/example.\n",
```
Comment (agsedova) [inline examples/trainer/simple_auto_trainer/notebook.ipynb:98]: "We have three splits: train, develop and test split. For each split, there is:
- a DataFrame, holding text (for training set) or text&labels (for development and test set)
- a Z matrix, relating instances, or rows in the DataFrame, to rules. 

Again, for more information we refer to the creation of the Dataset https://github.com/knodle/knodle/example.\n",

File: examples/trainer/simple_auto_trainer/notebook.ipynb [hunk 12]
```diff
@@ -0,0 +1,303 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "id": "included-reservation",
+   "metadata": {},
+   "source": [
+    "### Preprocess data to DistilBert input\n",
+    "\n",
+    "- Tokenize text. See https://huggingface.co/transformers/ on how to use Transformer-based models.\n",
+    "- Transform data to PyTorch."
```
Comment (agsedova) [inline examples/trainer/simple_auto_trainer/notebook.ipynb:143]: - Transform data to PyTorch-tensor format.

File: examples/trainer/simple_auto_trainer/notebook.ipynb [hunk 13]
```diff
@@ -0,0 +1,303 @@
...
+   "execution_count": null,
+   "id": "interracial-extent",
+   "metadata": {
+    "scrolled": true
+   },
+   "outputs": [],
+   "source": [
+    "from transformers import AutoModelForSequenceClassification, AdamW\n",
+    "\n",
+    "#from knodle.trainer import AutoTrainer, AutoConfig\n",
```
Comment (agsedova) [inline examples/trainer/simple_auto_trainer/notebook.ipynb:225]: delete commented lines

**Review by agsedova — CHANGES_REQUESTED** [review]
Have you seen my first review comments (about 1) having lists with pos and neg words in git repo; 2) explanation what autotrainer is)?
